crates/api_common/src/websocket/send.rs

```diff
@@ -23,253 +32,357 @@ use lemmy_db_schema::{
-pub async fn send_comment_ws_message_simple<OP: ToString + Send + OperationType + 'static>(
-  comment_id: CommentId,
-  op: OP,
-  context: &LemmyContext,
-) -> Result<CommentResponse, LemmyError> {
-  send_comment_ws_message(comment_id, op, None, None, None, vec![], context).await
-}
-
-#[tracing::instrument(skip_all)]
-pub async fn send_comment_ws_message<OP: ToString + Send + OperationType + 'static>(
-  comment_id: CommentId,
-  op: OP,
-  websocket_id: Option<ConnectionId>,
-  form_id: Option<String>,
-  person_id: Option<PersonId>,
-  recipient_ids: Vec<LocalUserId>,
-  context: &LemmyContext,
-) -> Result<CommentResponse, LemmyError> {
-  let view = CommentView::read(context.pool(), comment_id, person_id).await?;
-
-  let mut res = CommentResponse {
-    comment_view: view,
-    recipient_ids,
-    // The sent out form id should be null
-    form_id: None,
-  };
-
-  context
-    .chat_server()
-    .send_comment(&op, &res, websocket_id)
-    .await?;
-
-  // The recipient_ids should be empty for returns
-  res.recipient_ids = Vec::new();
-  res.form_id = form_id;
-
-  Ok(res)
-}
-
-#[tracing::instrument(skip_all)]
-pub async fn send_community_ws_message<OP: ToString + Send + OperationType + 'static>(
-  community_id: CommunityId,
-  op: OP,
-  websocket_id: Option<ConnectionId>,
-  person_id: Option<PersonId>,
-  context: &LemmyContext,
-) -> Result<CommunityResponse, LemmyError> {
-  let community_view =
-    CommunityView::read(context.pool(), community_id, person_id, Some(true)).await?;
-  let discussion_languages = CommunityLanguage::read(context.pool(), community_id).await?;
-
-  let mut res = CommunityResponse {
-    community_view,
-    discussion_languages,
-  };
-
-  // Strip out the person id and subscribed when sending to others
-  res.community_view.subscribed = SubscribedType::NotSubscribed;
-
-  context
-    .chat_server()
-    .send_community_room_message(&op, &res, res.community_view.community.id, websocket_id)
-    .await?;
-
-  Ok(res)
-}
-
-#[tracing::instrument(skip_all)]
-pub async fn send_pm_ws_message<OP: ToString + Send + OperationType + 'static>(
-  private_message_id: PrivateMessageId,
-  op: OP,
-  websocket_id: Option<ConnectionId>,
-  context: &LemmyContext,
-) -> Result<PrivateMessageResponse, LemmyError> {
-  let view = PrivateMessageView::read(context.pool(), private_message_id).await?;
-
-  let res = PrivateMessageResponse {
-    private_message_view: view,
-  };
-
-  // Send notifications to the local recipient, if one exists
-  if res.private_message_view.recipient.local {
-    let recipient_id = res.private_message_view.recipient.id;
-    let local_recipient = LocalUserView::read_person(context.pool(), recipient_id).await?;
-
-    context
-      .chat_server()
-      .send_user_room_message(&op, &res, local_recipient.local_user.id, websocket_id)
-      .await?;
+use serde::Serialize;
+
+impl LemmyContext {
+  #[tracing::instrument(skip_all)]
+  pub async fn send_post_ws_message(
```

Comment:
Good idea to move these methods into LemmyContext. But leave the generic parameter like before, its annoying to have to_string() all over the place.

Reply:
The generic op was way too complicated, mainly because for some reason the WebSocketOperation is split into 3 different enums. There was some hacks that I was able to remove by doing this.

Reply:
Which hacks? I think the generics are way cleaner on the call side, but maybe not in the websocket code itself.

Reply:
Another approach you can use is to take a generic param like `<OP: Into<String>`. Then you can treat it as string in the websocket code, while call sites dont have to put to_string() everywhere.

Reply:
I found a way to do it, I'll have this uploaded soon.